### ISSUE TITLE
Assumptions baked in the Graph optimizer not checked anywhere

### ISSUE THREAD

**Issue by qcolombet** [pytorch/glow#1178]
Some of the Graph optimizations assume that some nodes have variables as their inputs (e.g., batch normalization, convolution) and that these variables are used only in those nodes.
Although this currently matches what the importers emit, this creates an unwelcome coupling. In essence, we could spot usage of `cast<Variable>` sprinkled in the code base.
We should either:
1. Check these assumptions in the related verify methods, or
2. Fix the optimizer (and the hand written test cases)

I am tempted to go with #2 because:
A. Only the optimizer, importer and test cases have these assumptions baked in (for instance, the CPU optimization for convolution optimizeCPUConv is fine without those, same for the lowering and the interpreter.)
B. It is not unlikely that constant Variables with the same value would be uniqued and be shared across different uses and that's never checked.
C. With the introduce of the load node, those optimizations will need to be taught how to look through loads anyway.

**Comment by nadavrot** [comment]
I agree. The semantics should be that operators work on activations. Only special nodes such as save and load work on weights.

**Comment by nadavrot** [comment]
This is fixed now. Variables are gone, and save notes only operate on placeholders.
